instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/HistoryRecordingStateChangeObserverTest.kt

```diff
@@ -0,0 +1,508 @@
+    }
+
+    private fun generateRouteRefreshOptions(): RouteRefreshOptions {
+        val routeRefreshOptions = RouteRefreshOptions.Builder()
+            .intervalMillis(TimeUnit.SECONDS.toMillis(30))
+            .build()
+        RouteRefreshOptions::class.java.getDeclaredField("intervalMillis").apply {
+            isAccessible = true
+            set(routeRefreshOptions, 3_000L)
+        }
+        return routeRefreshOptions
+    }
+
+    private fun setUpMockRequestHandlersForRefresh(mockRoute: MockRoute) {
+        val failByRequestRouteRefreshResponse = FailByRequestMockRequestHandler(
+            MockDirectionsRefreshHandler(
+                mockRoute.routeResponse.uuid()!!,
+                readRawFileText(activity, R.raw.route_response_route_refresh_annotations)
+            )
+        )
+        mockWebServerRule.requestHandlers.add(failByRequestRouteRefreshResponse)
+        mockWebServerRule.requestHandlers.add(
+            MockRoutingTileEndpointErrorRequestHandler()
+        )
+    }
+
+    private fun setUpMockRequestHandlersForReroute(
+        mockRoute: MockRoute,
+        offRouteLocationUpdate: Location
+    ) {
+        mockWebServerRule.requestHandlers.addAll(mockRoute.mockRequestHandlers)
+        mockWebServerRule.requestHandlers.add(
+            MockDirectionsRequestHandler(
+                profile = DirectionsCriteria.PROFILE_DRIVING_TRAFFIC,
+                jsonResponse = readRawFileText(activity, R.raw.reroute_response_dc_very_short),
+                expectedCoordinates = listOf(
+                    Point.fromLngLat(
+                        offRouteLocationUpdate.longitude,
+                        offRouteLocationUpdate.latitude
+                    ),
+                    mockRoute.routeWaypoints.last()
+                ),
+            )
+        )
+    }
+
+    private fun getOffRouteLocation(originLocation: Point): Location =
+        mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = originLocation.latitude() + 0.002
+            longitude = originLocation.longitude()
+        }
+
+    private fun createMapboxNavigation() {
+        mapboxNavigation = MapboxNavigationProvider.create(
+            NavigationOptions.Builder(activity)
+                .accessToken(getMapboxAccessTokenFromResources(activity))
+                .build()
+        )
+    }
+
+    private fun observeHistoryRecordingEvents(
+        eventsChannel: Channel<HistoryRecordingStateChangeEvent>
+    ) {
+        val observer = object : HistoryRecordingStateChangeObserver {
+
+            override fun onShouldStartRecording(state: NavigationSessionState) {
+                CoroutineScope(Dispatchers.Main.immediate).launch {
+                    eventsChannel.send(
+                        HistoryRecordingStateChangeEvent(
+                            HistoryRecordingStateChangeEventType.START,
+                            state::class
+                        )
+                    )
+                }
```

Comment:
You can use `trySend` which is not `suspend`, if your channel has buffer. 
```suggestion
                eventsChannel.trySend(
                    HistoryRecordingStateChangeEvent(
                        HistoryRecordingStateChangeEventType.START,
                        state::class
                    )
                )
```
The same below. 
This creates a buffered channel:
```kotlin
val eventsChannel = Channel<HistoryRecordingStateChangeEvent>(Channel.BUFFERED)
```

Reply:
Cool, thanks!